chain/types/ethtypes/eth_transactions.go

```diff
@@ -0,0 +1,584 @@
+	}
+	return res, nil
+}
+
+func (tx *EthTxArgs) packSigFields() ([]interface{}, error) {
+	r, err := formatBigInt(tx.R)
+	if err != nil {
+		return nil, err
+	}
+
+	s, err := formatBigInt(tx.S)
+	if err != nil {
+		return nil, err
+	}
+
+	v, err := formatBigInt(tx.V)
+	if err != nil {
+		return nil, err
+	}
+
+	res := []interface{}{v, r, s}
+	return res, nil
+}
+
+func (tx *EthTxArgs) Signature() (*typescrypto.Signature, error) {
+	r := tx.R.Int.Bytes()
+	s := tx.S.Int.Bytes()
+	v := tx.V.Int.Bytes()
+
+	sig := append([]byte{}, padLeadingZeros(r, 32)...)
+	sig = append(sig, padLeadingZeros(s, 32)...)
+	if len(v) == 0 {
+		sig = append(sig, 0)
+	} else {
+		sig = append(sig, v[0])
+	}
+
+	if len(sig) != 65 {
+		return nil, fmt.Errorf("signature is not 65 bytes")
+	}
+	return &typescrypto.Signature{
+		Type: typescrypto.SigTypeDelegated, Data: sig,
+	}, nil
+}
+
+func (tx *EthTxArgs) Sender() (address.Address, error) {
+	msg, err := tx.ToRlpUnsignedMsg()
+	if err != nil {
+		return address.Undef, err
+	}
+
+	hasher := sha3.NewLegacyKeccak256()
+	hasher.Write(msg)
+	hash := hasher.Sum(nil)
+
+	sig, err := tx.Signature()
+	if err != nil {
+		return address.Undef, err
+	}
+
+	pubk, err := gocrypto.EcRecover(hash, sig.Data)
+	if err != nil {
+		return address.Undef, err
+	}
+
+	ethAddr, err := delegated.EthAddressFromPubKey(pubk)
+	if err != nil {
+		return address.Undef, err
+	}
+
+	return address.NewDelegatedAddress(builtintypes.EthereumAddressManagerActorID, ethAddr)
+}
+
+func RecoverSignature(sig typescrypto.Signature) (r, s, v EthBigInt, err error) {
+	if sig.Type != typescrypto.SigTypeDelegated {
+		return EthBigIntZero, EthBigIntZero, EthBigIntZero, fmt.Errorf("RecoverSignature only supports Delegated signature")
+	}
+
+	if len(sig.Data) != 65 {
+		return EthBigIntZero, EthBigIntZero, EthBigIntZero, fmt.Errorf("signature should be 65 bytes long, but got %d bytes", len(sig.Data))
+	}
+
+	r_, err := parseBigInt(sig.Data[0:32])
+	if err != nil {
+		return EthBigIntZero, EthBigIntZero, EthBigIntZero, fmt.Errorf("cannot parse r into EthBigInt")
+	}
+
+	s_, err := parseBigInt(sig.Data[32:64])
+	if err != nil {
+		return EthBigIntZero, EthBigIntZero, EthBigIntZero, fmt.Errorf("cannot parse s into EthBigInt")
+	}
+
+	v_, err := parseBigInt([]byte{sig.Data[64]})
+	if err != nil {
+		return EthBigIntZero, EthBigIntZero, EthBigIntZero, fmt.Errorf("cannot parse v into EthBigInt")
+	}
+
+	return EthBigInt(r_), EthBigInt(s_), EthBigInt(v_), nil
+}
+
+func parseEip1559Tx(data []byte) (*EthTxArgs, error) {
+	if data[0] != 2 {
+		return nil, fmt.Errorf("not an EIP-1559 transaction: first byte is not 2")
+	}
+
+	d, err := DecodeRLP(data[1:])
+	if err != nil {
+		return nil, err
+	}
+	decoded, ok := d.([]interface{})
+	if !ok {
+		return nil, fmt.Errorf("not an EIP-1559 transaction: decoded data is not a list")
+	}
+
+	if len(decoded) != 9 && len(decoded) != 12 {
+		return nil, fmt.Errorf("not an EIP-1559 transaction: should have 6 or 9 elements in the list")
```

Comment:
Given that the code below always accesses 12 elements of `decoded` shouldn't this be always checking for 12? Also the error message seems to have different numbers

Reply:
Yeah this seems off.

Reply:
Fixed.